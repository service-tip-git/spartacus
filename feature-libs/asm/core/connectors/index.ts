/*
 * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export * from './asm.adapter';
export * from './asm.connector';
export * from './converters';
