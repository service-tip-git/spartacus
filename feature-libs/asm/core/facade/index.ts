/*
 * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export * from './asm-create-customer.service';
export * from './asm-customer-list.service';
export * from './asm.service';
