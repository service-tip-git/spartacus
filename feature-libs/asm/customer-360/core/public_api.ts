/*
 * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export * from './connectors';
export * from './asm-customer-360-core.module';
export * from './services';
export * from './utils';
