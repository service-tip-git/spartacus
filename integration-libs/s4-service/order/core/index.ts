/*
 * SPDX-FileCopyrightText: 2025 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export * from './connector/index';
export * from './facade/index';
export * from './s4-service-order-core.module';
