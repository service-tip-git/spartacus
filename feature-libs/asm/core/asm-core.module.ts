/*
 * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MODULE_INITIALIZER } from '@spartacus/core';
import { AsmConnector } from './connectors/asm.connector';
import { facadeProviders } from './facade/facade-providers';
import { AsmStatePersistenceService } from './services/asm-state-persistence.service';
import { AsmStoreModule } from './store/asm-store.module';

export function asmStatePersistenceFactory(
  asmStatePersistenceService: AsmStatePersistenceService
): () => void {
  const result = () => asmStatePersistenceService.initSync();
  return result;
}

@NgModule({
  imports: [CommonModule, AsmStoreModule],
  providers: [
    AsmConnector,
    {
      provide: MODULE_INITIALIZER,
      useFactory: asmStatePersistenceFactory,
      deps: [AsmStatePersistenceService],
      multi: true,
    },
    ...facadeProviders,
  ],
})
export class AsmCoreModule {}
