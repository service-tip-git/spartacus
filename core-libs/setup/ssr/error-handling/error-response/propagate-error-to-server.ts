/*
 * SPDX-FileCopyrightText: 2025 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { InjectionToken } from '@angular/core';

/**
 * Propagates the given error object to the higher layer in the server.
 *
 * It's meant to propagate errors for example to ExpressJS layer when using SSR
 * or to a Prerendering Worker when using Server Prerendering.
 * Currently, it's provided OOTB only in SSR (not prerendering), in the `CxCommonEngine` class.
 *
 * Note: We need it until Angular implements a proper propagation of async errors
 * from an app to the the higher layer in the server.
 * For more, see the Angular issue https://github.com/angular/angular/issues/33642
 */
export const PROPAGATE_ERROR_TO_SERVER = new InjectionToken<
  (error: unknown) => void
>('PROPAGATE_ERROR_RESPONSE');
