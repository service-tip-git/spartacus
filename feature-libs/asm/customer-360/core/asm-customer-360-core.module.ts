/*
 * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { NgModule } from '@angular/core';
import { AsmCustomer360Connector } from './connectors';
import { facadeProviders } from './facade/facade-providers';
import { StoreFinderModule } from '@spartacus/storefinder';

@NgModule({
  imports: [StoreFinderModule],
  providers: [AsmCustomer360Connector, ...facadeProviders],
})
export class AsmCustomer360CoreModule {}
