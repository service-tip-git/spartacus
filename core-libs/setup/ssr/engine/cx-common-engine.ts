/*
 * SPDX-FileCopyrightText: 2025 SAP Spartacus team <spartacus-team@sap.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  CommonEngine,
  CommonEngineOptions,
  CommonEngineRenderOptions,
} from '@angular/ssr/node';
import { PROPAGATE_ERROR_TO_SERVER } from '../error-handling/error-response/propagate-error-to-server';

/**
 * The Spartacus extension of the Angular's `CommonEngine`. It is able to handle the propagated server responses caught during server-side rendering of a Spartacus app.
 * For reference, see Angular's source code: https://github.com/angular/angular-cli/blob/6cf866225ab09f8b4b3803c000b632bed8448ce4/packages/angular/ssr/src/common-engine.ts#L56
 *
 * @extends {CommonEngine}
 */
export class CxCommonEngine extends CommonEngine {
  constructor(options?: CommonEngineOptions) {
    super(options);
  }

  /**
   * @override
   * Renders for the given options.
   * If an error is populated from the rendered applications
   * (via `PROPAGATE_ERROR_TO_SERVER` callback), then such an error
   * will be thrown and the result promise rejected - but only AFTER the rendering is complete.
   * In other words, at first an error occurs and it's captured, then we wait until the rendering completes
   * and ONLY then we reject the promise with the payload being the encountered error.
   *
   * Note: if more errors are captured during the rendering, only the first one will be used
   *       as the payload of the rejected promise, others won't.
   *
   * @param {CommonEngineRenderOptions} options - The options to render.
   * @returns {Promise<string>} Promise which resolves with the rendered HTML as a string
   *                            OR rejects with the error, if any is propagated from the rendered app.
   */
  override async render(options: CommonEngineRenderOptions): Promise<string> {
    let error: undefined | unknown;

    return super
      .render({
        ...options,
        providers: [
          {
            provide: PROPAGATE_ERROR_TO_SERVER,
            useFactory: () => {
              return (propagatedError: unknown) => {
                // We're interested only the first propagated error, so we use `??=` instead of `=`:
                error ??= propagatedError;
              };
            },
          },
          ...(options.providers ?? []),
        ],
      })
      .then((html: string) => {
        if (error) {
          throw error;
        }
        return html;
      });
  }
}
