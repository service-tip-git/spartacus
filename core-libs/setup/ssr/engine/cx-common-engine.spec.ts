// eslint-disable-next-line import/no-unassigned-import
import '@angular/compiler';

import { Component, InjectionToken, NgModule, inject } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { ServerModule } from '@angular/platform-server';
import { PROPAGATE_ERROR_TO_SERVER } from '../error-handling/error-response/propagate-error-to-server';
import { CxCommonEngine } from './cx-common-engine';

// Test how the CxCommonEngine handles successful server-side rendering
@Component({
  selector: 'cx-mock',
  template: 'some template',
  standalone: false,
})
export class SuccessComponent {}

@NgModule({
  imports: [BrowserModule, ServerModule],
  declarations: [SuccessComponent],
  bootstrap: [SuccessComponent],
})
export class SuccessServerModule {}

// Test how the CxCommonEngine handles propagated error
@Component({
  selector: 'cx-response',
  template: ``,
  standalone: false,
})
export class WithPropagatedErrorComponent {
  constructor() {
    inject(PROPAGATE_ERROR_TO_SERVER)(new Error('test error'));
  }
}

@NgModule({
  imports: [BrowserModule, ServerModule],
  declarations: [WithPropagatedErrorComponent],
  bootstrap: [WithPropagatedErrorComponent],
})
export class WithPropagatedErrorServerModule {}

// Test that the CxCommonEngine doesn't override providers
// If SOME_TOKEN not provided, test how the CxCommonEngine handles APP_INITIALIZER errors
export const SOME_TOKEN = new InjectionToken<string>('SOME_TOKEN');

@Component({
  selector: 'cx-token',
  template: `message:{{ someToken }}`,
  standalone: false,
})
export class TokenComponent {
  someToken = inject(SOME_TOKEN);
}

@NgModule({
  imports: [BrowserModule, ServerModule],
  declarations: [TokenComponent],
  bootstrap: [TokenComponent],
})
export class TokenServerModule {}

describe('CxCommonEngine', () => {
  let engine: CxCommonEngine;

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation();
    jest.spyOn(console, 'log').mockImplementation();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should return html if no errors', async () => {
    engine = new CxCommonEngine({
      bootstrap: SuccessServerModule,
    });

    const html = await engine.render({
      url: 'http://localhost:4200',
      document: '<cx-mock></cx-mock>',
    });

    // Cannot use `.toMatchInlineSnapshot()` due to bug in jest:
    // see: https://github.com/thymikee/jest-preset-angular/issues/1084
    expect(html).toMatchSnapshot();
  });

  it('should not override providers passed to options', async () => {
    engine = new CxCommonEngine({
      bootstrap: TokenServerModule,
    });

    const html = await engine.render({
      url: 'http://localhost:4200',
      document: '<cx-token></cx-token>',
      providers: [{ provide: SOME_TOKEN, useValue: 'test' }],
    });

    // Cannot use `.toMatchInlineSnapshot()` due to bug in jest:
    // see: https://github.com/thymikee/jest-preset-angular/issues/1084
    expect(html).toMatchSnapshot();
  });

  it('should handle APP_INITIALIZER errors the standard Angular way and throw if any occurred', async () => {
    engine = new CxCommonEngine({
      bootstrap: TokenServerModule,
    });

    // Cannot use `.toMatchInlineSnapshot()` due to bug in jest:
    // see: https://github.com/thymikee/jest-preset-angular/issues/1084
    await expect(
      engine.render({
        url: 'http://localhost:4200',
        document: '<cx-token></cx-token>',
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should handle errors propagated from SSR', async () => {
    engine = new CxCommonEngine({
      bootstrap: WithPropagatedErrorServerModule,
    });

    // Cannot use `.toMatchInlineSnapshot()` due to bug in jest:
    // see: https://github.com/thymikee/jest-preset-angular/issues/1084
    await expect(
      engine.render({
        url: 'http://localhost:4200',
        document: '<cx-response></cx-response>',
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
