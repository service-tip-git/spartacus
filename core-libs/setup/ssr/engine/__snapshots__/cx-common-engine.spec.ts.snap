// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CxCommonEngine should handle APP_INITIALIZER errors the standard Angular way and throw if any occurred 1`] = `
"R3InjectorError(TokenServerModule)[InjectionToken SOME_TOKEN -> InjectionToken SOME_TOKEN]: 
  NullInjectorError: No provider for InjectionToken SOME_TOKEN!"
`;

exports[`CxCommonEngine should handle errors propagated from SSR 1`] = `"test error"`;

exports[`CxCommonEngine should not override providers passed to options 1`] = `"<html data-beasties-container><head></head><body><cx-token ng-version="19.0.3" ng-server-context="ssr">message:test</cx-token></body></html>"`;

exports[`CxCommonEngine should return html if no errors 1`] = `"<html data-beasties-container><head></head><body><cx-mock ng-version="19.0.3" ng-server-context="ssr">some template</cx-mock></body></html>"`;
